### PR TITLE
Fix ScreenManager not sending updates in certain cases that it should

### LINKED ISSUE #1536: Empty displayCapabilities in setDisplayLayout response prevents display update
### Bug Report
While processing the response to `setDisplayLayout`, the `sdl_displayLayoutResponse` method assumes the presence of `response.displayCapabilities` while the protocol specification indicates this is an optional field in the response. If the response contains an empty `displayCapabilities` field, then the cached value for `displayCapabilities` gets overwritten and subsequent attempts to update the display fail because the capabilities of the display cannot be determined. 

##### Reproduction Steps
1. Call `screenManager.setDisplayLayout`
2. Call `screenManager.beginUpdates()`
3. Call `screenManager.title = "Title"`
4. Call `screenManager.endUpdates()`

##### Expected Behavior
Title is shown on the screen

##### Observed Behavior
Title is not shown on the screen

##### OS & Version Information
* iOS Version: Simulator 11.4, iPhone 8+ 12.4
* SDL iOS Version: 6.4.1
* Testing Against: Emulator

##### Test Case, Sample Code, and / or Example App
```
smartDeviceLinkManager.sdlManager.send(request: setLayoutRequest) { _, response, error in
    guard response?.success.boolValue ?? false == true else {
        logError("Error setting display layout: \(String(describing: error))")
        return
    }
    let screenManager = self.smartDeviceLinkManager.sdlManager.screenManager
    screenManager.beginUpdates()
    screenManager.title = NSLocalizedString("My Title", comment: "")
    screenManager.primaryGraphic = self.logoArtwork
    screenManager.softButtonObjects = [self.meetingListButton, self.meetingRoomButton]
    screenManager.endUpdates()
}
```

### PR DESCRIPTION
**Pull request by joeljfischer**
Fixes #1536 

This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).

#### Unit Tests
- [x] Update unit tests to pass properly
- [x] Add new unit tests for sub-managers to ensure they send properly if `displayCapabilities` is `null`.

#### Core Tests
- [x] Tested the screen managers to ensure continued functionality.

Core version / branch / commit hash / module tested against: Sync 3.0 (17276_DEVTEST)
HMI name / version / branch / commit hash / module tested against: Sync 3.0 (17276_DEVTEST)

### Summary
This PR fixes a possible case where the head unit sends back a `nil` `displayCapabilities` or `displayCapabilities.textFields / imageFields` would cause the screen managers to not send data. Instead, we will now send all possible data and leave it to the head unit to reject or show what it determines is appropriate.

##### Enhancements
* Added a few RPC struct initializers.

##### Bug Fixes
* Fix a possible case where the head unit sends back a `nil` `displayCapabilities` or `displayCapabilities.textFields / imageFields` would cause the screen managers to not send data. Instead, we will now send all possible data and leave it to the head unit to reject or show what it determines is appropriate.
* Fixed the choice set manager sending data when the head unit sends `textFields` without the primary choice text.
* Fixed the soft button manager attempting to send soft buttons when no soft button capabilities are present.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: SmartDeviceLink/SDLDisplayCapability.h [hunk 1]
```diff
@@ -31,7 +31,7 @@ NS_ASSUME_NONNULL_BEGIN
  @param windowTypeSupported Informs the application how many windows the app is allowed to create per type.
  @param windowCapabilities Contains a list of capabilities of all windows related to the app. @see windowCapabilities
  */
-- (instancetype)initWithDisplayName:(NSString *)displayName windowTypeSupported:(nullable NSArray<SDLWindowCapability *> *)windowTypeSupported windowCapabilities:(nullable NSArray<SDLWindowTypeCapabilities *> *)windowCapabilities;
+- (instancetype)initWithDisplayName:(NSString *)displayName windowTypeSupported:(nullable NSArray<SDLWindowTypeCapabilities *> *)windowTypeSupported windowCapabilities:(nullable NSArray<SDLWindowCapability *> *)windowCapabilities;
```
[the remark this answers is not in the document]
Comment (joeljfischer) [reply SmartDeviceLink/SDLDisplayCapability.h]: Changed to not be breaking

File: SmartDeviceLink/SDLMenuManager.m [hunk 2]
```diff
@@ -197,6 +197,48 @@ - (void)setMenuCells:(NSArray<SDLMenuCell *> *)menuCells {
...
+- (BOOL)openMenu {
+    if ([SDLGlobals.sharedGlobals.rpcVersion isLessThanVersion:[[SDLVersion alloc] initWithMajor:6 minor:0 patch:0]]) {
+        SDLLogE(@"The openMenu method is not supported on this head unit.");
+        return NO;
+    }
+
+    SDLShowAppMenu *openMenu = [[SDLShowAppMenu alloc] init];
+
+    [self.connectionManager sendConnectionRequest:openMenu withResponseHandler:^(__kindof SDLRPCRequest * _Nullable request, __kindof SDLRPCResponse * _Nullable response, NSError * _Nullable error) {
+        if (error != nil) {
```
[the remark this answers is not in the document]
Comment (joeljfischer) [reply SmartDeviceLink/SDLMenuManager.m]: Enhanced logging

File: SmartDeviceLink/SDLMenuManager.m [hunk 3]
```diff
@@ -197,6 +197,48 @@ - (void)setMenuCells:(NSArray<SDLMenuCell *> *)menuCells {
...
+        return NO;
+    } else if (![self.menuCells containsObject:cell]) {
+        SDLLogE(@"This cell has not been sent to the head unit, so no submenu can be opened. Make sure that the cell exists in the SDLManager.menu array");
+        return NO;
+    }
+
+    SDLShowAppMenu *subMenu = [[SDLShowAppMenu alloc] initWithMenuID:cell.cellId];
+
+    [self.connectionManager sendConnectionRequest:subMenu withResponseHandler:^(__kindof SDLRPCRequest * _Nullable request, __kindof SDLRPCResponse * _Nullable response, NSError * _Nullable error) {
+        if (error != nil) {
```
[the remark this answers is not in the document]
Comment (joeljfischer) [reply SmartDeviceLink/SDLMenuManager.m]: Enhanced logging

File: SmartDeviceLink/SDLSoftButtonManager.m [hunk 4]
```diff
@@ -128,7 +139,8 @@ - (void)setSoftButtonObjects:(NSArray<SDLSoftButtonObject *> *)softButtonObjects
 
     _softButtonObjects = softButtonObjects;
 
-    SDLSoftButtonReplaceOperation *op = [[SDLSoftButtonReplaceOperation alloc] initWithConnectionManager:self.connectionManager fileManager:self.fileManager capabilities:self.windowCapability.softButtonCapabilities.firstObject softButtonObjects:_softButtonObjects mainField1:self.currentMainField1];
+    // We assume that all soft button capabilities are the same, so we only need to send one
+    SDLSoftButtonReplaceOperation *op = [[SDLSoftButtonReplaceOperation alloc] initWithConnectionManager:self.connectionManager fileManager:self.fileManager capabilities:self.windowCapability.softButtonCapabilities[0] softButtonObjects:_softButtonObjects mainField1:self.currentMainField1];
```
[the remark this answers is not in the document]
Comment (joeljfischer) [reply SmartDeviceLink/SDLSoftButtonManager.m]: Fixed

File: SmartDeviceLink/SDLSoftButtonManager.m [hunk 5]
```diff
@@ -194,14 +206,26 @@ - (void)setCurrentMainField1:(nullable NSString *)currentMainField1 {
...
 - (void)sdl_displayCapabilityDidUpdate:(SDLSystemCapability *)systemCapability {
-    self.windowCapability = self.systemCapabilityManager.defaultMainWindowCapability;
+    NSArray<SDLDisplayCapability *> *capabilities = systemCapability.displayCapabilities;
+    if (capabilities == nil || capabilities.count == 0) {
+        self.windowCapability = nil;
+    } else {
+        SDLDisplayCapability *mainDisplay = capabilities[0];
+        for (SDLWindowCapability *windowCapability in mainDisplay.windowCapabilities) {
+            NSUInteger currentWindowID = windowCapability.windowID != nil ? windowCapability.windowID.unsignedIntegerValue : SDLPredefinedWindowsDefaultWindow;
+            if (currentWindowID == SDLPredefinedWindowsDefaultWindow) {
```
[the remark this answers is not in the document]
Comment (joeljfischer) [reply SmartDeviceLink/SDLSoftButtonManager.m]: The if statement you note should have a break, which I have added. But I'm confused by your comment "avoid another nested if statement." Could you explain what you mean?

File: SmartDeviceLink/SDLTextField.h [hunk 6]
```diff
@@ -48,6 +48,13 @@ NS_ASSUME_NONNULL_BEGIN
  */
 @property (strong, nonatomic) NSNumber<SDLInt> *rows;
 
+/// Convenience initalizer for the TextField RPC struct
+/// @param name The name identifying this text field
+/// @param characterSet The character set of this text field
+/// @param width The number of characters per row allowed in this text field
+/// @param rows The number of rows allowed in this text field, deliniated by a newline character "\n"
```
[the remark this answers is not in the document]
Comment (joeljfischer) [reply SmartDeviceLink/SDLTextField.h]: Removed. I don't know where I got that but it's not in the spec. IDK what rows means...

File: SmartDeviceLink/SDLChoiceSetManager.m [hunk 7]
```diff
@@ -139,6 +145,19 @@ - (NSOperationQueue *)sdl_newTransactionQueue {
     return queue;
 }
 
+- (void)sdl_updateTransactionQueueSuspended {
```
[the remark this answers is not in the document]
Comment (joeljfischer) [reply SmartDeviceLink/SDLChoiceSetManager.m]: I did a bunch of testing, and it appears that on Sync 3 at least, it does work in ALERT or HMI_OBSCURED states, so I'm removing those checks and adding documentation for the rest

File: SmartDeviceLink/SDLChoiceSetManager.m [hunk 8]
```diff
@@ -36,8 +36,10 @@
 #import "SDLRPCResponseNotification.h"
 #import "SDLSetDisplayLayoutResponse.h"
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLChoiceSetManager.m]: The `SDLCheckChoiceVROptionalOperation` does not get sent when the `hmiLevel` is non-`NONE` the first time I connect via USB cord. However, if I pull and reconnect the USB cord, then the `SDLCheckChoiceVROptionalOperation` gets sent when the `hmiLevel` is still `NONE`.

File: SmartDeviceLink/SDLChoiceSetManager.m [hunk 9]
```diff
@@ -139,11 +144,23 @@ - (NSOperationQueue *)sdl_newTransactionQueue {
     return queue;
 }
 
+/// Suspend the transaction queue if we are in HMI NONE
+/// OR if we don't know what text fields are allowed, or if the text field name "menu name" (this is the primary choice text) we assume we cannot present a PI.
```
[the remark this answers is not in the document]
Comment (joeljfischer) [reply SmartDeviceLink/SDLChoiceSetManager.m]: Resolved

File: SmartDeviceLink/SDLChoiceSetManager.m [hunk 10]
```diff
@@ -433,35 +473,34 @@ - (NSString *)currentState {
...
+    NSArray<SDLDisplayCapability *> *capabilities = systemCapability.displayCapabilities;
+    if (capabilities == nil || capabilities.count == 0) {
+        self.currentWindowCapability = nil;
+    } else {
+        SDLDisplayCapability *mainDisplay = capabilities[0];
+        for (SDLWindowCapability *windowCapability in mainDisplay.windowCapabilities) {
+            NSUInteger currentWindowID = windowCapability.windowID != nil ? windowCapability.windowID.unsignedIntegerValue : SDLPredefinedWindowsDefaultWindow;
+            if (currentWindowID == SDLPredefinedWindowsDefaultWindow) {
+                self.currentWindowCapability = windowCapability;
+            }
```
[the remark this answers is not in the document]
Comment (joeljfischer) [reply SmartDeviceLink/SDLChoiceSetManager.m]: Fixed with an extra newline

File: SmartDeviceLink/SDLSoftButtonManager.m [hunk 11]
```diff
@@ -194,15 +208,34 @@ - (void)setCurrentMainField1:(nullable NSString *)currentMainField1 {
...
+    NSArray<SDLDisplayCapability *> *capabilities = systemCapability.displayCapabilities;
+    if (capabilities == nil || capabilities.count == 0) {
+        self.softButtonCapabilities = nil;
+    } else {
+        SDLDisplayCapability *mainDisplay = capabilities[0];
+        for (SDLWindowCapability *windowCapability in mainDisplay.windowCapabilities) {
+            NSUInteger currentWindowID = windowCapability.windowID != nil ? windowCapability.windowID.unsignedIntegerValue : SDLPredefinedWindowsDefaultWindow;
+            if (currentWindowID == SDLPredefinedWindowsDefaultWindow) {
+                self.softButtonCapabilities = windowCapability.softButtonCapabilities.firstObject;
+                break;
```
[the remark this answers is not in the document]
Comment (joeljfischer) [reply SmartDeviceLink/SDLSoftButtonManager.m]: Resolved with an extra newline

File: SmartDeviceLink/SDLChoiceSetManager.m (hunk 8 above)
Comment (joeljfischer) [reply SmartDeviceLink/SDLChoiceSetManager.m]: Fixed; the HMI level wasn't getting reset properly
